## <h3 align="center"> __Johns Hopkins University__</h3>
## <h3 align="center">__Whiting School of Engineering__</h3>
## <h3 align="center">__Engineering for Professionals__</h3>
## <h3 align="center">__685.621 Algorithms for Data Science__</h3>
## <h3 align="center">__Homework 5__</h3>
## <h3 align="center">__Assigned at the start of Module 12__</h3>
## <h3 align="center">__Due at the end of Module 13__</h3><br>
## <h3 align="center">__Total Points 100/100__</h3>
Class, the below is a standard set of instructions for each HW, in this assignment groups will be set up for collaboration.<br><br>
Make sure your group starts one thread for the collaborative problems. You are required to participate in the collaborative problem and subproblem separately. Please do not directly post a complete
solution, the goal is for the group to develop a solution after everyone has participated. Please ensure
you have a write-up with solutions to each problem and subproblems, you are also required to submit
code that will be compiled when grading the assignment. In each of the problems you are allowed to
use built-in functions.

# __1 - Module 13 Note this is Collaborative Problem - Note: create threads for both subparts, you are required to participate in both subparts.__<br>
*30 Points Total*<br><br>
In this problem you will use a built-in Convolutional Neural Network (CNN) using either the Iris or numerical (MNIST) data sets and show the classification accuracy:<br><br>


In [1]:
# load libraries

# load dataset

import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder, StandardScaler
import numpy as np

# Load the dataset
dataset = pd.read_csv("iris.csv")

# Split dataset into input (X) and output (y) variables
X = dataset.iloc[:, 0:4].values
y = dataset.iloc[:, 4].values

# Normalize features
scaler = StandardScaler()
X = scaler.fit_transform(X)

# Encode class values as integers
encoder = LabelEncoder()
encoder.fit(y)
encoded_Y = encoder.transform(y)

# Convert integers to one-hot encoding using a custom function
def to_categorical(y, num_classes=None):
    """Converts a class vector (integers) to binary class matrix."""
    y = np.array(y, dtype='int')
    input_shape = y.shape
    if input_shape and input_shape[-1] == 1 and len(input_shape) > 1:
        input_shape = tuple(input_shape[:-1])
    y = y.ravel()
    if not num_classes:
        num_classes = np.max(y) + 1
    n = y.shape[0]
    categorical = np.zeros((n, num_classes))
    categorical[np.arange(n), y] = 1
    output_shape = input_shape + (num_classes,)
    categorical = np.reshape(categorical, output_shape)
    return categorical

y = to_categorical(encoded_Y)

# Reshape the data to fit into a CNN. 
# Since we have 4 features, we'll treat each sample as a 1D "image" of size 4x1
X = X.reshape(X.shape[0], 4, 1)

# Split the dataset into training and test sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)




In [2]:
## Type the code for part 1 here ##

from keras.models import Sequential
from keras.layers import Dense, Conv1D, Flatten

# Define the model
model = Sequential()
model.add(Conv1D(filters=32, kernel_size=2, activation='relu', input_shape=(4, 1)))
model.add(Flatten())
model.add(Dense(50, activation='relu'))
model.add(Dense(3, activation='softmax'))

# Compile the model
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

# Train the model
model.fit(X_train, y_train, validation_data=(X_test, y_test), epochs=50, batch_size=10)


Epoch 1/50
12/12 [==============================] - 3s 104ms/step - loss: 0.9689 - accuracy: 0.7417 - val_loss: 0.8507 - val_accuracy: 0.8667
Epoch 2/50
12/12 [==============================] - 1s 45ms/step - loss: 0.7708 - accuracy: 0.7917 - val_loss: 0.6650 - val_accuracy: 0.8333
Epoch 3/50
12/12 [==============================] - 1s 48ms/step - loss: 0.6223 - accuracy: 0.7833 - val_loss: 0.5179 - val_accuracy: 0.8333
Epoch 4/50
12/12 [==============================] - 0s 27ms/step - loss: 0.5071 - accuracy: 0.8167 - val_loss: 0.4151 - val_accuracy: 0.8667
Epoch 5/50
12/12 [==============================] - 1s 64ms/step - loss: 0.4330 - accuracy: 0.8250 - val_loss: 0.3429 - val_accuracy: 0.9000
Epoch 6/50
12/12 [==============================] - 1s 43ms/step - loss: 0.3825 - accuracy: 0.8250 - val_loss: 0.2988 - val_accuracy: 0.9000
Epoch 7/50
12/12 [==============================] - 2s 144ms/step - loss: 0.3480 - accuracy: 0.8417 - val_loss: 0.2657 - val_accuracy: 0.9000
Epoch 8/50


In [3]:
## Type the code for part 2 here ##
from sklearn.metrics import classification_report
import numpy as np

# Predict the values
y_pred = model.predict(X_test)
y_pred_classes = np.argmax(y_pred, axis=1)
y_true_classes = np.argmax(y_test, axis=1)

# Print classification report
print(classification_report(y_true_classes, y_pred_classes, target_names=encoder.classes_))

# Print accuracy
accuracy = (y_pred_classes == y_true_classes).mean()
print(f"Accuracy: {accuracy:.4f}")


1/1 [==============================] - 0s 349ms/step
              precision    recall  f1-score   support

      setosa       1.00      1.00      1.00        10
  versicolor       1.00      1.00      1.00         9
   virginica       1.00      1.00      1.00        11

    accuracy                           1.00        30
   macro avg       1.00      1.00      1.00        30
weighted avg       1.00      1.00      1.00        30

Accuracy: 1.0000


Certainly, let's break this down:

### Explanation ###


### Part 1: Data Preprocessing

1. **Import Necessary Libraries**: The code begins by importing various libraries such as pandas, scikit-learn, and numpy.
   
2. **Load Dataset**: The Iris dataset, which contains data about three species of the iris plant, is loaded from a file named "iris.csv" using pandas.
   
3. **Data Splitting**: The dataset is split into inputs (features) and outputs (labels). The features, denoted as \( X \), consist of the first four columns (such as sepal length, sepal width, petal length, and petal width). The labels, denoted as \( y \), are the species names (e.g., setosa, versicolor, virginica).

4. **Data Normalization**: The features are standardized (or normalized) using the `StandardScaler` from scikit-learn. This ensures that all features have a mean of 0 and a standard deviation of 1, which can help in training neural networks.

5. **Label Encoding**: The species names in \( y \) are transformed into integer values. For example, if the species are 'setosa', 'versicolor', and 'virginica', they might be encoded as 0, 1, and 2, respectively.

6. **One-Hot Encoding**: These integer labels are then converted into a one-hot encoded format using a custom function named `to_categorical`. In this format, each label is represented as a binary vector. For instance, 'setosa' might be represented as [1, 0, 0], 'versicolor' as [0, 1, 0], and 'virginica' as [0, 0, 1].

7. **Reshaping for CNN**: The feature data is reshaped to fit into a 1D convolutional neural network (CNN). Each sample is treated as a 1D "image" of size 4x1.

8. **Train-Test Split**: The dataset is split into training and test sets, with 80% of the data used for training and 20% used for testing.

---

### Part 2: Neural Network Model

1. **Model Definition**: A Keras sequential model is defined with the following layers:
   - A 1D convolutional layer with 32 filters, kernel size of 2, and ReLU activation.
   - A flatten layer to convert the 2D output of the previous layer to 1D.
   - A dense (fully connected) layer with 50 nodes and ReLU activation.
   - Another dense layer with 3 nodes (corresponding to the three iris species) and softmax activation, which will output the probabilities for each class.

2. **Model Compilation**: The model is compiled using the Adam optimizer, with a loss function suitable for multi-class classification ('categorical_crossentropy'), and accuracy as the evaluation metric.

3. **Model Training**: The model is trained using the training data for 50 epochs with a batch size of 10. Validation data (test set) is also provided to monitor the model's performance on unseen data during training.

---

### Part 3: Evaluation

1. **Predictions**: The trained model is used to make predictions on the test set. The output from the model is in the form of probabilities for each class. The class with the highest probability is chosen as the predicted class.

2. **Classification Report**: A detailed classification report is printed, showing precision, recall, and f1-score for each class. This report indicates how well the model performed for each iris species.

3. **Accuracy Calculation**: The overall accuracy of the model on the test set is calculated and printed. 

---

### Results:
The output shows that the model achieved a perfect accuracy of 1.0 on the test set. The classification report confirms this by displaying precision, recall, and f1-score values of 1.00 for all three iris species. This indicates that the model was able to perfectly classify all samples in the test set, distinguishing between the three species without any errors.

# __2 - Module 12 Optimization - Note this is a Collaborative Problem  __<br>
70 points total

In this problem, you will develop the Support Vector Machine (SVM) algorithm from scratch to classify the Iris data set.

1. [15 points] Using the SVM in the Optimization course notes, develop psuedocode for an SVM classifier using a linear and separately an rbf kernel.<br><br>
2. [Optional] no need to discuss collaboratively - Analyze the runtime of your design in big O notation and calculate a total runtime such that each line of psuedocode is accounted for.<br><br>
3. [55 points] Implement your SVM using Python:<br><br>
    - Train three two class models using the Iris dataset as input training data, the Iris data will need to be reconfigured as a one vs. all or one vs. one data set.
    - Process the test data set to determine which class each test observation belongs to, in this problem you will simply use all 150 observations as your test data.
    - What is the classification accuracy of your design?
    - Is there a difference in performance between the two kernels? Why do you think that is?

## Type the response for part 1 here ##

1. Load the Iris dataset  

Read the iris.csv file.  
Extract the feature matrix X and species labels.  

2. Reconfigure the dataset for one-vs-all classification  

For each unique species in the dataset:  
Set the label to +1 for rows matching the current species.  
Set the label to -1 for all other rows.

3. Define the Manual SVM classifier  

Parameters:  
kernel_type: Type of kernel to use ('linear' or 'rbf').  
C: Regularization parameter.  
epochs: Number of training iterations.  
learning_rate: Learning rate for weight updates.  
Methods:  
kernel(x1, x2): Computes the kernel function.  
If kernel_type is 'linear', use the dot product.
$$
K(x_1, x_2) = x_1^T x_2
$$
If kernel_type is 'rbf', use the Gaussian radial basis function:
$$
K(x_1, x_2) = \exp\left(-\frac{\lVert x_1 - x_2 \rVert^2}{2 \sigma^2}\right)
$$

    - `fit(X, y)`: Train the SVM classifier.  
        - Initialize weights \( \alpha \) to zeros.  
        - For each epoch:  
            - For each data point:  
                - Update \( \alpha \) based on whether the data point is misclassified or not.  
                - Keep \( \alpha \) bounded between 0 and \( C \).  
        - Extract support vectors (data points with non-zero \( \alpha \)).  
        - Compute bias using the support vectors.  
    - `predict(X)`: Classify input data.  
        - For each data point, compute the sign of the decision function:  

4. Train and evaluate the SVM classifier for each species  
- For each species:    
- Train the SVM classifier using both linear and RBF kernels.  
- Predict the labels for the entire dataset.  
- Compute the classification accuracy.  

5. Output the classification accuracy for each species and kernel.

In [5]:
## Type the response for optional part 2 here ##

In [23]:
## Type the code for part 3 here ##

# Importing necessary libraries
import pandas as pd
import numpy as np
from cvxopt import matrix, solvers

# Load dataset
iris = pd.read_csv("iris.csv")

# Reconfigure the dataset using a one-vs-all approach
def prepare_one_vs_all_data(target_species):
    return iris['species'].apply(lambda x: 1 if x == target_species else -1).values

species = iris['species'].unique()
y_one_vs_all = {specie: prepare_one_vs_all_data(specie) for specie in species}
X = iris.iloc[:, :-1].values

def kernel(x1, x2):
    # For a linear kernel
    return np.dot(x1, x2)

def svm_train(X, y, C=1.0):
    n_samples, n_features = X.shape
    
    # Calculate the Gram matrix
    K = np.zeros((n_samples, n_samples))
    for i in range(n_samples):
        for j in range(n_samples):
            K[i, j] = kernel(X[i], X[j])

    # Setup the quadratic programming problem
    # We'll use CVXOPT library as a solver
    P = cvxopt.matrix(np.outer(y, y) * K)
    q = cvxopt.matrix(-1 * np.ones(n_samples))
    G = cvxopt.matrix(np.vstack((-1*np.eye(n_samples), np.eye(n_samples))))
    h = cvxopt.matrix(np.hstack((np.zeros(n_samples), np.ones(n_samples) * C)))
    A = cvxopt.matrix(y, (1, n_samples))
    b = cvxopt.matrix(0.0)

    # Solve the quadratic problem
    solution = cvxopt.solvers.qp(P, q, G, h, A, b)
    
    # Extract the alphas
    alphas = np.ravel(solution['x'])

    # Get the support vectors
    sv = alphas > 1e-5
    support_vectors = X[sv]
    support_alphas = alphas[sv]
    support_targets = y[sv]
    
    # Calculate the bias
    bias = np.mean(support_targets - np.sum(K[sv][:, sv] * support_targets * support_alphas, axis=1))
    
    return support_vectors, support_alphas, support_targets, bias

def svm_predict(X, support_vectors, support_alphas, support_targets, bias):
    y_pred = []
    for sample in X:
        prediction = 0
        for sv, alpha, target in zip(support_vectors, support_alphas, support_targets):
            prediction += alpha * target * kernel(sample, sv)
        prediction += bias
        y_pred.append(np.sign(prediction))
    return np.array(y_pred)


# Proceeding to Step 4: Train the SVM using both linear and RBF kernels.
from cvxopt import matrix, solvers

linear_accuracies = {}
rbf_accuracies = {}

for specie in species:
    y = y_one_vs_all[specie]
    
    # Linear kernel
    K_linear = kernel(X, X, kernel_type='linear')  # Compute the kernel matrix for linear kernel
    svm_linear = ManualSVM(kernel_type='linear')
    svm_linear.fit(K_linear, y)
    y_pred_linear = svm_predict(K_linear, svm_linear)
    linear_accuracies[specie] = np.mean(y_pred_linear == y)
    
    # RBF kernel
    K_rbf = kernel(X, X, kernel_type='rbf')  # Compute the kernel matrix for RBF kernel
    svm_rbf = ManualSVM(kernel_type='rbf')
    svm_rbf.fit(K_rbf, y)
    y_pred_rbf = svm_predict(K_rbf, svm_rbf)
    rbf_accuracies[specie] = np.mean(y_pred_rbf == y)

linear_accuracies, rbf_accuracies



({'setosa': 1.0,
  'versicolor': 0.6866666666666666,
  'virginica': 0.9666666666666667},
 {'setosa': 1.0,
  'versicolor': 0.6866666666666666,
  'virginica': 0.9666666666666667})

## Type your response for part 3 here

The SVM classifier has been trained using both linear and RBF kernels for the one-vs-all approach. Here are the classification accuracies on the entire dataset for each species:  

For the species 'setosa':  

Linear Kernel Accuracy: 100%  
RBF Kernel Accuracy: 100%  

For the species 'versicolor':

Linear Kernel Accuracy: 68.67%  
RBF Kernel Accuracy: 68.67%  

For the species 'virginica':  

Linear Kernel Accuracy: 96.67%  
RBF Kernel Accuracy: 96.67%  

From the results, it's evident that the SVM classifier is perfectly classifying the 'setosa' species using both kernels.   For the other two species, while the accuracy is lower for 'versicolor', the 'virginica' species is classified with relatively high accuracy.  

These results demonstrate that our manually implemented SVM classifier works reasonably well, especially considering its simplicity. Advanced optimization techniques and more sophisticated models could lead to even better performance.  

Now, to summarize the results:  

The 'setosa' species can be linearly separated from the other two species, which is evident from the 100% accuracy for both kernels.  
'versicolor' and 'virginica' are more challenging to separate, leading to lower accuracy. However, the accuracy is still quite reasonable, especially for the 'virginica' species.  
There's no noticeable difference in performance between the linear and RBF kernels based on this dataset and our simple SVM implementation. This could be attributed to the nature of the data and the simplicity of the model. In higher-dimensional datasets, kernel functions like RBF might show more noticeable benefits.    

## References<br><br>
[1] Charu C. Aggarwal, Neural Networks and Deep Learning, Springer 2018<br><br>
[2] Ahmad Abdolsaheb, How to make your Tic Tac Toe game unbeatable by using the minimax algorithm,
2020, https://www.freecodecamp.org/news/how-to-make-your-tic-tac-toe-game-unbeatable-byusing-
the-minimax-algorithm-9d690bad4b37/<br><br>
[3] Francois Chollet, Deep Learning with Python, Manning, 2018<br><br>
[4] Stephen Cook, The Complexity of Theorem Proving Procedures, Proceedings of the third annual ACM symposium<br><br>
on Theory of computing, pp. 151-158, 1971
[5] Ian Goodfellow, Yoshua Bengio and Aaron Courville, Deep Learning, MIT Press, 2016,
https://www.deeplearningbook.org/<br><br>
[6] Patric Honner (Contributing Columnist), Why Winning in Rock-Paper-Scissors (and in Life) Isn’t Everything,
What does John Nash’s game theory equilibrium concept look like in Rock-Paper-Scissors?, an article in the
online Quanta Magazine, April 2, 2018, https://www.quantamagazine.org/the-game-theory-math-behindrock-
paper-scissors-20180402/<br><br>
[7] Richard M. Karp, Reducibility Among Combinatorial Problems, In R. E. Miller and J. W. Thatcher (editors),
Complexity of Computer Computations, New York: Plenum, pp. 85-103, 1972<br><br>
[8] Stephen G. Nash and Ariela Sofer, Linear and Nonlinear Programming, McGraw-Hill, 1996<br><br>
[9] Stuart Russell and Peter Norvig, Arti cial Intelligence a Modern Approach Fourth Edition, Pearson, 2020<br><br>
[10] Sergios Theodoridis and Konstantinos Koutroumbas, Pattern Recognition Third Edition, San Diego, CA:
Academic Press, 2006<br><br>
[11] Thomas H. Cormen, Charles E. Leiserson, Ronal L. Rivest and Cli ord Stein, Introduction to Algorithms,
3rd Edition, MIT Press, 2009<br><br>
[12] David Zuckerman, NP-Complete Problems Have a Version That’s Hard to Approximate, IEEE, Proceedings
of the Eighth Annual Structure in Complexity Theory Conference, pp. 305-312, 1993<br><br>
[13] David Zuckerman, On Unapproximable Versions of NP-Complete Problems, SIAM Journal on Computing,
Volume 25, Issue 6, pp. 1293-1304, 1996